lib/elixir/lib/exception.ex

```diff
@@ -924,14 +924,34 @@ defmodule UndefinedFunctionError do
 
   defp message(:"function not exported", module, function, arity) do
     formatted_fun = Exception.format_mfa(module, function, arity)
-    {"function #{formatted_fun} is undefined or private", true}
+    fun_message = "function #{formatted_fun} is undefined or private"
+    behaviour_hint = behaviour_hint(module, function, arity)
+    {fun_message <> behaviour_hint, true}
   end
 
   defp message(reason, module, function, arity) do
     formatted_fun = Exception.format_mfa(module, function, arity)
     {"function #{formatted_fun} is undefined (#{reason})", false}
   end
 
+  defp behaviour_hint(module, function, arity) do
+    case behaviours_for(module) do
+      [] ->
+        ""
+
+      behaviours ->
+        case Enum.find(behaviours, &expects_callback?(&1, function, arity)) do
+          nil -> ""
+          behaviour -> ", but the behaviour #{inspect(behaviour)} expects it to be present"
+        end
+    end
+  end
+
+  defp expects_callback?(behaviour, function, arity) do
+    callbacks = behaviour.behaviour_info(:callbacks)
+    Enum.member?(callbacks, {function, arity})
```

Comment:
We also need to make sure the callback is not in the optional callbacks. In a nutshell:

```elixir
pair = {function, arity}
pair in behaviour.behaviour_info(:callbacks) and pair not in behaviour.behaviour_info(:optional_callbacks)
```

Reply:
Good call, let's keep it as is then!